soub blog post draft (#1721)
informalsystems/quint: docs/content/posts/soup.md

## TODO: This needs better Presentation (Arrow to represent the transformation)
```
pure def upon_vote_msg(vote: VoteMsg, s: LocalState): Transition =
    if (s.is_leader() and s.is_valid(vote))
        val s1 = s.add_vote(vote)
        if (s1.get_votes(vote.id).size() >= Q)
            val qc = s1.build_qc(vote.id)
            val p =  s1.build_proposal(qc)
            {postState: s1 , effects: Set(BroadcastProp(p))}
        else
            nop(s1)
    else
        nop(s)
```

```
pure def upon_vote_quorum(vote_soup: Set[VoteMsg], s: LocalState): Set[Transition] =
    if (s.is_leader())
        vote_soup.filter(v => s.is_valid(v))
                 .groupBy(v => v.id)
                 .values()
                 .filter(q => q.size() >= Q)
                 .map(vs => s.build_qc(vs))
                 .map(qc => s.build_proposal(qc))
                 .map(p => {postState: s , effects: Set(BroadcastProp(p))})
    else
        Set(nop(s))
```

> Now let's compare the two MonadBFT specifications: the original with explicit bookkeeping versus our refactored version using the message soup pattern. To compare the performance of the two approaches, we use a set of witnesses that explore different scenarios ranging from a simple reproposal case to more complex scenarios like the disappearance of a high tip from subsequent view. A witness is something that we expect to hold in at least one state in one execution. We can then ask the Quint simulator whether it can find a trace that leads to a state that is described by the witness. We conducted 2 sets of experiments that answer these questions: How fast can a model find a specific witness "once"? What is the frequency of witness observation across multiple samples?

Author: ```suggestion
Now let's compare the two MonadBFT specifications: the original with explicit bookkeeping versus our refactored version using the message soup pattern. To compare the performance of the two approaches, we use a set of witnesses. A witness is something that we expect to hold in at least one state in one execution. We can then ask the Quint simulator whether it can find a trace that leads to a state that is described by the witness. We conducted 2 sets of experiments that answer these questions: How fast can a model find a specific witness "once"? What is the frequency of witness observation across multiple samples?

The witnesses we use in our experiments explore different scenarios. Some are quite easy to construct in short executions (like 
a simple case were a proposal is reproposed) to more complex scenarios that require longer executions to reach them (like the disappearance of a high tip from subsequent view). 
```
"the disappearance of a high tip from subsequent view" is a quite big jump and requires knowing the algorithm. I wonder whether we should say more here, or e.g., just give a pointer to the paper

Whether you're manually analyzing traces or feeding them to an LLM to parse and explain the behavior, shorter and more coarse-grained traces are always better. For engineers, it's about cognitive load as you can actually follow the consensus flow without drowning in message-handling mechanics. For language models, it's about token efficiency—why burn through your context window on message tracking when you could focus those tokens on the actual algorithmic logic?

## TODO: ADD BAR PLOT HERE FOR THE EXAMPLE ABOVE

## A Need for Speed
The message soup version finds witnesses at least 3x faster than the bookkeeping approach, and that's in the worst case. For more complex scenarios, the message-per-message version would often time out or run out of memory entirely while the message soup version sailed through. This speed matters because faster executions mean faster debugging cycles and more extensive coverage of the state space. When you can explore more scenarios in the same amount of time, you're establishing more trust in your system. More speed, more trust, more sleep.

## TODO: ADD REST OF BAR PLOTS HERE

> ## What Doesn't Kill You Makes You Stronger

Author: ```suggestion
## When the Soup Gets Spicy: Byzantine Faults
```

> A powerful trick to model byzantine behavior is to simply pre-populate the message soup with all messages that byzantine nodes could ever send without modeling complex byzantine node logic. The message soup approach handles the load that comes with this technique more elegantly than explicit bookkeeping: because our system only makes transitions on enabled steps, having a huge number of possibilities for byzantine messages doesn't cause state space explosion. In the bookkeeping approach, every byzantine message creates a new state transition that the system has to explore, leading to exponential blowup. 

Author: ```suggestion
A powerful trick to model byzantine behavior is to simply pre-populate the message soup with all messages that byzantine nodes could ever send without modeling complex byzantine node logic. (We have used this in our earlier [blog post on Alpenglow](https://quint-lang.org/posts/alpenglow), where we wanted to analyze the issue of having too many faults.)
The message soup approach handles the load that comes with this technique more elegantly than explicit bookkeeping: because our system only makes transitions on enabled steps, having a huge number of possibilities for byzantine messages doesn't cause state space explosion. In the bookkeeping approach, every byzantine message creates a new state transition that the system has to explore, leading to exponential blowup. 
```
It is not only about "state space explosion". It also about more branching and longer executions. Should we discuss this here?

Author: OK. I have changed "state space explosion" to "combinatorial explosion" in the proposed text. It is a bit less specific, which is good here I guess. But we don't need to add more discussion; I agree. So if it is fine for you @beu5a , please just accept the change...

The message soup approach handles the load that comes with this technique more elegantly than explicit bookkeeping: because our system only makes transitions on enabled steps, having a huge number of possibilities for byzantine messages doesn't cause combinatorial explosion. In the bookkeeping approach, every byzantine message creates a new state transition that the system has to explore, leading to exponential blowup. 
The plot below shows resilience to byzantine messages injection. We can see that the frequency of observing witnesses is unaffected for the message soup approach, but the bookkeeping version fails to maintain this resilience. In the single-message version, there's a massive gap between the Byzantine and honest scenarios. Byzantine conditions severely degrade performance while the message soup approach shows barely any difference.
## TODO: ADD LINE PLOT HERE
When you're dealing with adversarial networks, what doesn't kill your protocol makes it stronger. The message soup cuts through the byzantine noise and helps you focus on what really matters: is your design still safe when under attack?

List of blog posts/resources for consensus in Quint:

> - [Alpenglow](https://quint-lang.org/posts/alpenglow)

Author: ```suggestion
- [Solana Alpenglow](https://quint-lang.org/posts/alpenglow)
```

## Long Story Short
Remember how we said the message soup approach skips intermediate states? Well, this shows up dramatically in the traces. The same consensus scenario where a reproposal is observed that takes on average **37 steps with the message soup needed 500+ steps in the original bookkeeping version**. That's not just a numbers game, it's the difference between seeing the consensus logic clearly versus getting lost in translation, when looking at the output Quint generates.
Whether you're manually analyzing traces or feeding them to an LLM to parse and explain the behavior, shorter and more coarse-grained traces are always better. For engineers, it's about cognitive load as you can actually follow the consensus flow without drowning in message-handling mechanics. For language models, it's about token efficiency—why burn through your context window on message tracking when you could focus those tokens on the actual algorithmic logic?

## TODO: ADD BAR PLOT HERE FOR THE EXAMPLE ABOVE

## The Need for Speed

> The message soup version finds witnesses at **least 3x faster** than the bookkeeping approach, and that's in the worst case. For more complex scenarios, the message-per-message version would often time out or run out of memory entirely while the message soup version sailed through. This speed matters because faster executions mean faster debugging cycles and more extensive coverage of the state space. When you can explore more scenarios in the same amount of time, you're establishing more trust in your system. More speed, more trust, more sleep.

Reviewer: ```suggestion
The message soup version finds witnesses **at least 3x faster** than the bookkeeping approach, and that's in the worst case. For more complex scenarios, the message-per-message version would often time out or run out of memory entirely while the message soup version sailed through. This speed matters because faster executions mean faster debugging cycles and more extensive coverage of the state space. When you can explore more scenarios in the same amount of time, you're establishing more trust in your system. More speed, more trust, more sleep.
```

The message soup approach handles the load that comes with this technique more elegantly than explicit bookkeeping: because our system only makes transitions on enabled steps, having a huge number of possibilities for byzantine messages doesn't cause combinatorial explosion. In the bookkeeping approach, every byzantine message creates a new state transition that the system has to explore, leading to exponential blowup. 
The plot below shows resilience to byzantine messages injection. We can see that the frequency of observing witnesses is unaffected for the message soup approach, but the bookkeeping version fails to maintain this resilience. In the single-message version, there's a massive gap between the Byzantine and honest scenarios. Byzantine conditions severely degrade performance while the message soup approach shows barely any difference.
## TODO: ADD LINE PLOT HERE
When you're dealing with adversarial networks, what doesn't kill your protocol makes it stronger. The message soup cuts through the byzantine noise and helps you focus on what really matters: is your design still safe when under attack?

List of blog posts/resources for consensus in Quint:

> - [Solana Alpenglow](https://quint-lang.org/posts/alpenglow)
> - [Espresso Hotshot](https://informal.systems/blog/espresso-hotshot-epoch-changes-in-quint-2025)
> - [ZKSync ChonkyBFT](https://protocols-made-fun.com/consensus/matterlabs/quint/specification/modelchecking/2024/07/29/chonkybft.html)
> - [Commonware Minimmit](https://github.com/commonwarexyz/monorepo/tree/main/pipeline/minimmit/quint)

Author: ```suggestion
- Quint blog post on [Solana Alpenglow](https://quint-lang.org/posts/alpenglow)
- Informal's blog post on [Espresso Hotshot](https://informal.systems/blog/espresso-hotshot-epoch-changes-in-quint-2025)
- Igor Konnov's blog post on [ZKSync ChonkyBFT](https://protocols-made-fun.com/consensus/matterlabs/quint/specification/modelchecking/2024/07/29/chonkybft.html)
- Commonware's Quint spec on [Minimmit](https://github.com/commonwarexyz/monorepo/tree/main/pipeline/minimmit/quint)
```
- The [MonadBFT spec](https://github.com/informalsystems/monad-spec) discussed in this post

The message soup approach handles the load that comes with this technique more elegantly than explicit bookkeeping: because our system only makes transitions on enabled steps, having a huge number of possibilities for byzantine messages doesn't cause combinatorial explosion. In the bookkeeping approach, every byzantine message creates a new state transition that the system has to explore, leading to exponential blowup. 
The plot below shows resilience to byzantine messages injection. We can see that the frequency of observing witnesses is unaffected for the message soup approach, but the bookkeeping version fails to maintain this resilience. In the single-message version, there's a massive gap between the Byzantine and honest scenarios. Byzantine conditions severely degrade performance while the message soup approach shows barely any difference.
## TODO: ADD LINE PLOT HERE
When you're dealing with adversarial networks, what doesn't kill your protocol makes it stronger. The message soup cuts through the byzantine noise and helps you focus on what really matters: is your design still safe when under attack?

> List of blog posts/resources for consensus in Quint:
> - Quint blog post on [Solana Alpenglow](https://quint-lang.org/posts/alpenglow)
> - Informal's blog post on [Espresso Hotshot](https://informal.systems/blog/espresso-hotshot-epoch-changes-in-quint-2025)
> - Igor Konnov's blog post on [ZKSync ChonkyBFT](https://protocols-made-fun.com/consensus/matterlabs/quint/specification/modelchecking/2024/07/29/chonkybft.html)
> - Commonware's Quint spec on [Minimmit](https://github.com/commonwarexyz/monorepo/tree/main/pipeline/minimmit/quint)
> - The [MonadBFT spec](https://github.com/informalsystems/monad-spec) discussed in this post

Author: ```suggestion
List of blog posts/resources for consensus in Quint:
- The [MonadBFT spec](https://github.com/informalsystems/monad-spec) discussed in this post
- Quint blog post on [Solana Alpenglow](https://quint-lang.org/posts/alpenglow)
- Informal's blog post on [Espresso Hotshot](https://informal.systems/blog/espresso-hotshot-epoch-changes-in-quint-2025)
- Igor Konnov's blog post on [ZKSync ChonkyBFT](https://protocols-made-fun.com/consensus/matterlabs/quint/specification/modelchecking/2024/07/29/chonkybft.html)
- Commonware's Quint spec on [Minimmit](https://github.com/commonwarexyz/monorepo/tree/main/pipeline/minimmit/quint)
- The examples in the Quint repo contain a [text book consensus algorithm](https://github.com/informalsystems/quint/blob/main/examples/classic/distributed/ConsensusAlgorithm/ConsensusAlg.qnt) and [Tendermint](https://github.com/informalsystems/quint/tree/main/examples/cosmos/tendermint)
```
